Crashlytics/Crashlytics/Models/FIRCLSOnDemandModel.m

```diff
@@ -0,0 +1,209 @@
+    // is stored until send or delete unsent reports is called.
+    [self incrementQueuedOperationCount:1];
+    [self getOrIncrementOnDemandEventCountForCurrentRun:YES];
+    dispatch_async(self.dispatchQueue, ^{
+      if (dataCollectionEnabled) {
+        [existingReportManager handleOnDemandReportUpload:activeReportPath
+                                      dataCollectionToken:dataCollectionToken
+                                                 asUrgent:YES];
+      } else {
+        [self.storedActiveReportPaths insertObject:activeReportPath atIndex:0];
+        if ([self.storedActiveReportPaths count] > FIRCLSMaxUnsentReports) {
+          [self.fileManager removeItemAtPath:[self.storedActiveReportPaths lastObject] error:nil];
+          [self.storedActiveReportPaths removeLastObject];
+          self->_recordedOnDemandExceptionCount -= 1;
+          self->_droppedOnDemandExceptionCount += 1;
+        }
+      }
+      double uploadDelay = [self calculateUploadDelay];
+      if (dataCollectionEnabled) {
+        FIRCLSDebugLog(@"Submitted an on-demand exception, starting delay %.20f", uploadDelay);
+      } else {
+        FIRCLSDebugLog(@"Stored an on-demand exception, starting delay %.20f", uploadDelay);
+      }
+      sleep(uploadDelay);
+      [self incrementQueuedOperationCount:-1];
+    });
+    return YES;  // Recorded and submitted the exception.
+  }
+}
+
+// Helper methods
+
+- (double)calculateUploadDelay {
+  double calculatedStepDuration = [self calculateStepDuration];
+  double power = pow(self.base, calculatedStepDuration);
+  NSNumber *calculatedUploadDelay =
+      [NSNumber numberWithDouble:(SEC_PER_MINUTE / self.uploadRate) * power];
+  NSComparisonResult result =
+      [[NSNumber numberWithDouble:MAX_DELAY_SEC] compare:calculatedUploadDelay];
+  return (result == NSOrderedAscending) ? MAX_DELAY_SEC : [calculatedUploadDelay doubleValue];
+}
+
+- (double)calculateStepDuration {
+  double currentTime = [NSDate timeIntervalSinceReferenceDate];
+  BOOL queueIsFull = [self isQueueFull];
+
+  if (self.currentStep == -1) {
+    self.currentStep = 0;
+    self.lastUpdated = currentTime;
+  }
+
+  double delta = (currentTime - self.lastUpdated) / (double)self.stepDuration;
+  double queueFullDuration = (self.currentStep + delta) > 100 ? 100 : (self.currentStep + delta);
+  double queueNotFullDuration = (self.currentStep - delta) < 0 ? 0 : (self.currentStep - delta);
+  double calculatedDuration = queueIsFull ? queueFullDuration : queueNotFullDuration;
+
+  if (self.currentStep != calculatedDuration) {
+    self.currentStep = calculatedDuration;
+    self.lastUpdated = currentTime;
+  }
+
+  return calculatedDuration;
+}
+
+- (int)getOrIncrementOnDemandEventCountForCurrentRun:(BOOL)increment {
+  @synchronized(self) {
+    if (increment) {
+      _recordedOnDemandExceptionCount += 1;
+    }
+    return self.recordedOnDemandExceptionCount;
+  }
+}
+
+- (int)getOrIncrementDroppedOnDemandEventCountForCurrentRun:(BOOL)increment {
```

Comment:
I think we can call these 2 methods `increment...` and drop the `getOr...` - think it's clear enough what it returns

Reply:
Wait nvm